src/sentry_plugins/bitbucket/plugin.py

```diff
@@ -205,3 +238,137 @@ def view_autocomplete(self, request, group, **kwargs):
+            try:
+                repo = client.get_repo(config['name'])
+            except Exception as e:
+                self.raise_error(e)
+            else:
+                config['external_id'] = six.text_type(repo['uuid'])
+        return config
+
+    def get_webhook_secret(self, organization):
+        lock = locks.get('bitbucket:webhook-secret:{}'.format(organization.id),
+                         duration=60)
+        with lock.acquire():
+            secret = OrganizationOption.objects.get_value(
+                organization=organization,
+                key='bitbucket:webhook_secret',
+            )
+            if secret is None:
+                secret = uuid4().hex + uuid4().hex
+                OrganizationOption.objects.set_value(
+                    organization=organization,
+                    key='bitbucket:webhook_secret',
+                    value=secret,
+                )
+        return secret
+
+    def create_repository(self, organization, data, actor=None):
+        if actor is None:
+            raise NotImplementedError('Cannot create a repository anonymously')
+
+        client = self.get_client(actor)
+        try:
+            resp = client.create_hook(data['name'], {
+                'description': 'sentry-bitbucket-repo-hook',
+                'url': absolute_uri('/plugins/bitbucket/organizations/{}/webhook/'.format(organization.id)),
+                'active': True,
+                'events': ['repo:push'],
+            })
+        except Exception as e:
+            self.raise_error(e)
+        else:
+            return {
+                'name': data['name'],
+                'external_id': data['external_id'],
+                'url': 'https://bitbucket.org/{}'.format(data['name']),
+                'config': {
+                    'name': data['name'],
+                    'webhook_id': resp['uuid'],
+                }
+            }
+
+    def delete_repository(self, repo, actor=None):
+        if actor is None:
+            raise NotImplementedError('Cannot delete a repository anonymously')
+
+        client = self.get_client(actor)
+        try:
+            client.delete_hook(repo.config['name'], repo.config['webhook_id'])
+        except ApiError as exc:
+            if exc.code == 404:
+                return
+            raise
+
+    def _format_commits(self, repo, commit_list):
+        return [{
+            'id': c['hash'],
+            'repository': repo.name,
+            'author_email': parse_raw_user(c['author']['raw']),
+            'author_name': c['author']['user']['display_name'],
+            'message': c['message'],
+            'patch_set': c.get('patch_set'),
+        } for c in commit_list]
+
+    def compare_commits(self, repo, start_sha, end_sha, actor=None):
+        if actor is None:
+            raise NotImplementedError('Cannot fetch commits anonymously')
+
+        client = self.get_client(actor)
+        # use config name because that is kept in sync via webhooks
```

Comment:
now it https://github.com/getsentry/sentry-plugins/pull/180/files#diff-746655ad5cb98464cff47451391017e2R66 here now